README.md
# svg2png

[![Build Status](https://travis-ci.org/cdaniel/svg2png.svg?branch=master)](https://travis-ci.org/cdaniel/svg2png)

A service that converts svg to png


## Run the app locally

1. [Install Node.js][]
4. Run `npm install` to install the app's dependencies
5. Run `npm start` to start the app
6. Access the running app in a browser at http://localhost:6001

[Install Node.js]: https://nodejs.org/en/download/

CHANGELOG.md
## Aug 5, 2016
- Basic working functionality

## Feb 2, 2016
- Change engine to Node v4.2
- Change express version to 4.13

## Nov 16, 2015
- Added changelog


manual_test.sh
curl -X POST -F file=@tests/_image.svg -o r2.png http://svg-to-png.eu-gb.mybluemix.net\?width=600\&height=400


tests/tests.js
//#!/bin/env node
/* Copyright 2016 Leading Edge Forum

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

var should = require('should');
var app = require('../lib/app');
var request = require('supertest');

describe('POST /', function() {

    var _assertErrorMessage = function(msg) {
        return function(res) {
            var errorMsg = JSON.parse(res.text).error;
            if (errorMsg !== msg) {
                throw new Error('Unexpected error message:' + errorMsg +',\n while it should be ' + msg);
            }
        }
    };
    
    it('should complain about invalid width', function(done) {
        request(app)
        .post('/')
        .expect(422)
        .expect(_assertErrorMessage("Width(undefined) should be > 100 but lower than 2000"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });

    it('should complain about too small width', function(done) {
        request(app)
        .post('/?width=5')
        .expect(422)
        .expect(_assertErrorMessage("Width(5) should be > 100 but lower than 2000"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });

    it('should complain about too big width', function(done) {
        request(app)
        .post('/?width=5000')
        .expect(422)
        .expect(_assertErrorMessage("Width(5000) should be > 100 but lower than 2000"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });
    
    it('should complain about missing heigth', function(done) {
        request(app)
        .post('/?width=500')
        .expect(422)
        .expect(_assertErrorMessage("Height(undefined) should be > 80 but lower than 1200"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });
    
    it('should complain about heigth being too small', function(done) {
        request(app)
        .post('/?width=500&height=10')
        .expect(422)
        .expect(_assertErrorMessage("Height(10) should be > 80 but lower than 1200"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });
    
    it('should complain about heigth being too large', function(done) {
        request(app)
        .post('/?width=500&height=2000')
        .expect(422)
        .expect(_assertErrorMessage("Height(2000) should be > 80 but lower than 1200"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });
    
    it('should complain about missing file', function(done) {
        request(app)
        .post('/?width=500&height=250')
        .expect(422)
        .expect(_assertErrorMessage("No file received"))
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });
    
    it('should generate a file', function(done) {
        request(app)
        .post('/?width=500&height=250')
        .attach('file','tests/_image.svg')
        .expect(200)
        .end(function(err, res) {
            if (err) throw err;
            done();
        });
    });

});


lib/app.js
//#!/bin/env node
/* Copyright 2016 Leading Edge Forum

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

// This application uses express as its web server
// for more info, see: http://expressjs.com
var express = require('express');

// cfenv provides access to your Cloud Foundry environment
// for more info, see: https://www.npmjs.com/package/cfenv
var cfenv = require('cfenv');

// create a new express server
var app = express();

app.use(require('express-xml-bodyparser')());

// serve the files out of ./public as our main files
app.use('/', require('./svg2png.js')().router);

// get the app environment from Cloud Foundry
var appEnv = cfenv.getAppEnv();

// start server on the specified port and binding host
module.exports = app.listen(appEnv.port, '0.0.0.0', function() {
  // print a message when the server starts listening
  console.log("server starting on " + appEnv.url);
});


lib/svg2png.js
//#!/bin/env node
/* Copyright 2016 Leading Edge Forum

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

var atob = require('atob');
var phantom = require('phantom');
var multer = require('multer');


var upload = multer({
    storage: multer.memoryStorage(),
    limits : {
        fileSize : 1024*1024*1024,
        files : 1
    }
    });

var svgToPNG = function(svgXML, width, height, res){

  phantom.create(['--ignore-ssl-errors=yes']).then(function(ph) {

      ph.createPage().then(function(page) {

          page.property('viewportSize', {
              width : width,
              height : height
          });

          page.property('content',
          '<?xml version="1.0" encoding="UTF-8" standalone="no"?>' + svgXML);

          page.renderBase64('png').then(function(content64){
            res.writeHead(200, {
                'Content-Type': 'image/png'
            });
            res.end(atob(content64),'binary');
          });

          page.close();
          ph.exit();

      }).catch(
        function(e){
          console.log(e);
          res.end();
        }
      );
  }).catch(
    function(e){
      console.log(e);
      res.statusCode = 500;
      res.end();
    }
  );
};


module.exports = function(){
    var module = {};


    module.router = require('express').Router();

    // do actual conversion
    module.router.post('/' , upload.single('file'), function(req, res) {
        //TODO: sanity checks
        var width = req.query.width;
        var height = req.query.height;
        if(!width || width < 100 || width > 2000){
            res.status(422).json({ error: 'Width(' + width + ') should be > 100 but lower than 2000' });
            return;
        }
        if(!height || height < 100 || height > 1200){
            res.status(422).json({ error: 'Height(' + height + ') should be > 80 but lower than 1200' });
            return;
        }
        var svgFile = req.file; //unparsed
        if(!svgFile){
            res.status(422).json({ error: 'No file received' });
            return;
        }
        
        var svgXML = svgFile.buffer.toString('UTF-8');

        if(!svgXML.length > 1024 * 1024 * 1024 /*1MB*/){
            res.status(422).json({ error: 'The file is too big' });
            return;
        }
        svgToPNG(svgXML, width, height, res);
    });


    return module;
};

